app/grandchallenge/algorithms/views.py

```diff
@@ -1013,3 +1022,52 @@ def get_success_message(self, cleaned_data):
 
     def get_success_url(self):
         return self.algorithm.get_absolute_url() + "#models"
+
+
+class AlgorithmImageTemplate(ObjectPermissionRequiredMixin, DetailView):
+    model = Algorithm
+    permission_required = "algorithms.change_algorithm"
+    raise_exception = True
+    queryset = Algorithm.objects.prefetch_related("inputs", "outputs")
+
+    def get(self, *_, **__):
+        algorithm = self.get_object()
+
+        forge_context = get_forge_algorithm_template_context(algorithm)
+
+        # Step 1: Create a temporary directory for generating content
+        with tempfile.TemporaryDirectory() as temp_dir:
+            # Step 2: Generate the content in the temp directory
+            algorithm_template_path = generate_algorithm_template(
+                context=forge_context,
+                output_path=Path(temp_dir),
+            )
+
+            # Step 3: Create an in-memory buffer to store the ZIP file
+            zip_buffer = io.BytesIO()
+
+            # Step 4: Create the ZIP archive in memory
+            with ZipFile(zip_buffer, "w") as zipf:
+                for foldername, _, filenames in os.walk(
+                    algorithm_template_path
+                ):
+                    for filename in filenames:
+                        file_path = os.path.join(foldername, filename)
```

Comment:
Django has a function called `safe_join` that you can use in order to ensure that a file is within a certain directory.

Reply:
safe_join is under django.utils._os ... not sure if the _ is indicative of internal use. Given the documentation incorrectly mentions a raised ValueError I've taken the precaution of implementing the check myself but have taken over the exception it raises. Imho it's applicable.

Reply:
Yes fine for us to use it, we've been using it for years and I think the API is pretty stable.